src/library/lastplayedcache.cpp

```diff
@@ -0,0 +1,74 @@
+#include "library/lastplayedcache.h"
+
+#include <QSqlDatabase>
+#include <QtDebug>
+
+#include "library/dao/playlistdao.h"
+#include "library/queryutil.h"
+#include "library/trackcollection.h"
+
+LastPlayedCache::LastPlayedCache(TrackCollection* trackCollection)
+        : m_pTrackCollection(trackCollection) {
+    initTableView();
+
+    connect(&m_pTrackCollection->getPlaylistDAO(),
+            &PlaylistDAO::trackAdded,
+            this,
+            &LastPlayedCache::slotPlaylistTrackChanged);
+    connect(&m_pTrackCollection->getPlaylistDAO(),
+            &PlaylistDAO::trackRemoved,
+            this,
+            &LastPlayedCache::slotPlaylistTrackChanged);
+}
+
+void LastPlayedCache::initTableView() {
+    QSqlQuery lastPlayedQuery(m_pTrackCollection->database());
+    // Views can't use params, so just use .args here.  There is no injection
+    // risk because these are constants.
+    lastPlayedQuery.prepare(QStringLiteral(
+            "CREATE TEMPORARY VIEW IF NOT EXISTS %1 AS "
+            "SELECT "
+            "  PlaylistTracks.track_id, "
+            "  MAX(PlaylistTracks.pl_datetime_added) as datetime_played "
+            "FROM PlaylistTracks "
+            "JOIN Playlists ON PlaylistTracks.playlist_id == Playlists.id "
+            "WHERE Playlists.hidden = %2 "
+            "GROUP BY PlaylistTracks.track_id")
+                                    .arg(LASTPLAYEDTABLE_NAME,
+                                            QString::number(PlaylistDAO::
+                                                            PLHT_SET_LOG)));
+    if (!lastPlayedQuery.exec()) {
+        LOG_FAILED_QUERY(lastPlayedQuery);
+    }
+}
+
+void LastPlayedCache::slotPlaylistTrackChanged(
+        int playlistId, TrackId trackId, int /* a_iPosition */) {
+    if (m_pTrackCollection->getPlaylistDAO().getHiddenType(playlistId) !=
+            PlaylistDAO::PLHT_SET_LOG) {
+        return;
+    }
+
+    TrackPointer pTrack = m_pTrackCollection->getTrackById(trackId);
+    pTrack->setLastPlayedDate(fetchLastPlayedTime(m_pTrackCollection->database(), pTrack));
+}
+
+// static, so that trackdao.cpp can fetch values for initial population in the cache.
+QDateTime LastPlayedCache::fetchLastPlayedTime(const QSqlDatabase& db, TrackPointer pTrack) {
+    QSqlQuery updateQuery(db);
```

Comment:
The name is misleading, this is not an update.

Reply:
Could we prepare the query once and store it as a member? We don't need to prepare the same query over and over again. This only works for the same database connection, i.e. the method would need to be non-static.

Reply:
I would like it to be non-static, but trackdao doesn't have access to the lastplayedcache object created by mixxxlibraryfeature.  I could make lastplayedcache something that can be instantiated multiple times but only if there's a nice way to make sure the setup isn't done twice.  For the DB view, I can add IF NOT EXISTS, but for the qt connections I'm not sure how best to do that.   (reference counting?)

Reply:
Also, trackdao has access to a database connection, but not a trackcollection

Reply:
(I could just duplicate the query in trackdao but the idea was to have all the lastplayedcache functions in one file instead of spreading the logic all over mixxx)

Reply:
fixed